### PR TITLE
Refactor MiniBatch and stop using deprecated MRG sampler

### LINKED ISSUE #4523: VI using deprecated Aesara MRG sampler
The current usage of the mrng sampler in ADVI does not support JAX (https://github.com/pymc-devs/aesara/issues/322). It should be fairly easy to instead make it use the new `RandomVariable` Op https://github.com/pymc-devs/aesara/pull/296 to get JAX support for ADVI.

CC @ferrine

### LINKED ISSUE #6277: Use RandomVariables for Minibatch
Minibatch creates a stochastic view of the underlying dataset using some pretty funky Aesara magic. 

In a straightforward implementation it would do something like:
```python
import pymc as pm
import numpy as np

data = pm.Normal.dist(size=(100, 2)).eval()

with pm.Model() as m:
  data = pm.Data("data", data)

  mb_slice_start, mb_slice_end = pm.Uniform.dist(0, data.shape[0], size=(2)).sort()
  minibatch_data = data[mb_slice_start: mb_slice_end]
  
  x = pm.Normal("x", 0, 1, observed=minibatch_data)
```
And then PyMC also figures out that `x` lopg has to be scaled by the size of the unobserved values in each iteration.

One pressing issues is that it relies on the deprecated MRG sampler #4523, which also has no support for JAX or Numba. It isn't as simple a matter as switching to the default `RandomStream` which returns `RandomVariable`s, because we treat the shared RNG variables used by RandomVariables in a special manner inside `aesaraf.compile_pymc` (the function through which all PyMC functions go through before calling `aesara.function`). Specifically we always set the values of the distinct shared RNG variables to something new, based on what `random_seed` is passed to `compile_pymc`.

This is a problem when you need to synchronize multiple mini-batches of data, say when you have a linear regression model with a minibatch of `x` and `y`. The way it currently works is that you set the same `random_seed` to each (default is hard-coded 42), and then rely on this initial value being the same so that the endpoints of the slice stay synchronized. But if we were to use the standard RandomStream, and pass the graph to `compile_pymc`, the two RNGs associated with the two slices would be overwritten to different values, regardless of the fact that they started with the same value.

Anyway, this trick might also not have been enough, since there was a need to introduce `align_minibatches` in https://github.com/pymc-devs/pymc/pull/2760: 
https://github.com/pymc-devs/pymc/blob/faebc60ff2cb7c9befff5fc1122894ffa5600139/pymc/data.py#L463-L474

https://github.com/pymc-devs/pymc/blob/2296350959e4035b4f1ee13fab88014d4f0fa545/pymc/tests/test_data.py#L754-L773

The big issue here is that we are not representing that the slice endpoints are the same symbolically. The current graph looks something like:

```python
with pm.Model() as m:
  x = pm.Data("x", x)
  y = pm.Data("y", y)

  rng = aesara.shared(np.random.default_rng(42))
  mb_slice_start, mb_slice_end = pm.Uniform.dist(0, data.shape[0], size=(2), rng=rng).sort()
  minibatch_x = x[mb_slice_start: mb_slice_end]

  rng = aesara.shared(np.random.default_rng(42))
  mb_slice_start, mb_slice_end = pm.Uniform.dist(0, data.shape[0], size=(2), rng=rng).sort()
  minibatch_y = y[mb_slice_start: mb_slice_end]
  
  obs = pm.Normal("obs", x, 1, observed=minibatch_y)
```
The two rngs were set to the same initial value, but other than that they are completely different as far as Aesara cares. The correct graph would be:

```python
with pm.Model() as m:
  x = pm.Data("x", x)
  y = pm.Data("y", y)
  
  rng = aesara.shared(np.random.default_rng(42))
  mb_slice_start, mb_slice_end = pm.Uniform.dist(0, data.shape[0], shape=(2,), rng=rng).sort()
  minibatch_x = x[mb_slice_start: mb_slice_end]
  minibatch_y = y[mb_slice_start: mb_slice_end]
  
  obs = pm.Normal("obs", x, 1, observed=minibatch_y)
```

That is, using the same slicing RV for both mini-batches. Regardless of the seed value that is set by `compile_pymc`, the two minibatches will always be compatible.

## The incremental proposal

Refactor Minibatch so that it can accept multiple variables at once, which will share the same random slices. Then start using RandomVariables instead of MRG stuff, and get rid of seed at the `minibatch` level

```python
with pm.Model() as m:
  x, y = pm.Minibatch({x: (10, ...), y: (10)})
```
Maybe even sprinkle some dims if you want

## The radical proposal
Offer the same API, but implement the minibatch view with the more straightforward code from the pseudo-examples in this issue. 

No need to keep a global list of all the RNGs:

https://github.com/pymc-devs/pymc/blob/faebc60ff2cb7c9befff5fc1122894ffa5600139/pymc/data.py#L304

The whole Op view magic shouldn't be needed AFAICT. The only thing is that right now PyMC complains if you pass something to observed other than a shared variable or constant data. The concern here is that the observations should not depend on other model RVs, but that's easy to check with `aesara.graph.ancestors()`

https://github.com/pymc-devs/pymc/blob/faebc60ff2cb7c9befff5fc1122894ffa5600139/pymc/data.py#L339

The other issue is that we don't allow orphan RVs in the logp graph, but we can easily create a subtype like `MinibatchUniformRV`, that is allowed, just like `SimulatorRV`s are allowed.

This should make the Minibatch code much more readable and maintainable, as well as compatible with the non-C backends (I have no idea if either supports `View`)

### PR DESCRIPTION
**Pull request by ferrine**
fix #4523, suspend #6273
closes #6277 

@ricardoV94, I've figured out the root cause of the seeding bug

### CODE REVIEW

**Comment by ricardoV94** [comment]
@ferrine this does not fix #4523, VI is still using the MRG RandomStream. This can't be switched yet because of #6277

**Comment by ricardoV94** [comment]
It does fix some lack of seeding, for which I think we don't have an issue open

File: pymc/variational/opvi.py
Comment (ricardoV94) [inline pymc/variational/opvi.py:387]: Specifically `random_seed` in `compile_pymc` won't do anything with `MRG` `RNG` state variables. We don't actually want to do that until we sort out #6277, otherwise two distinct minibatch RNGs would always be given different initial seed values.

Comment (ferrine) [reply pymc/variational/opvi.py:387]: hmm

Comment (ferrine) [reply pymc/variational/opvi.py:387]: Once minibatching is fixed, it should become valid, right?

Comment (ferrine) [reply pymc/variational/opvi.py:387]: Does it make sense to fix in one PR and get rid of MRG?

Comment (ricardoV94) [reply pymc/variational/opvi.py:387]: Yes it does!

File: pymc/data.py
Comment (ricardoV94) [inline pymc/data.py]: We can also get rid of `at_rng`, as the seeds created by it will be overwritten in `compile_pymc` anyway. Just import `RandomStream` wherever it's needed.

File: pymc/model.py
Comment (ricardoV94) [inline pymc/model.py]: You're not using a custom TensorType anymore? If so you can check against that.

Comment (ferrine) [reply pymc/model.py]: no custom tensor type(

File: pymc/data.py
Comment (ferrine) [reply pymc/data.py]: ok

File: pymc/model.py
Comment (ricardoV94) [reply pymc/model.py]: I think `_joint_logp` will complain when it sees the RVs in the logp graph? Have you not seen that?

Comment (ricardoV94) [reply pymc/model.py]: If you see that, the easiest workaround would be to subclass the Uniform RV used in the Minibatch Slice and allow that type here:

https://github.com/pymc-devs/pymc/blob/5da26173428cacc6aa5e1d05e4b51ceb69dadfd9/pymc/distributions/logprob.py#L107-L126

`and not isinstance(node.owner.op, (SimulatorRV, MinibatchSliceUniformRV))`

Comment (ferrine) [reply pymc/model.py]: I've added tests that only pass after I implement minibatching with a new RV

File: pymc/tests/test_data.py
Comment (ricardoV94) [inline pymc/tests/test_data.py]: `joint_logp` is deprecated. You can test that it does not fail by just calling `m.logp()`

**Comment by ricardoV94** [comment]
I think I saw that kernel crash in my PR, perhaps an out of bounds indexing?

**Comment by ferrine** [comment]
no idea why this windows test fails

**Comment by ferrine** [comment]
hmm, that could be the random stream!

**Comment by ferrine** [comment]
The issue that blocks the PR seems to be in aesara or I'm doing something wrong https://github.com/aesara-devs/aesara/issues/1301

**Comment by ferrine** [comment]
Maybe @brandonwillard or @ricardoV94 have any insights?

**Comment by michaelosthege** [comment]
@ferrine are some of these changes independent and could be merged via a separate PR already?

**Comment by ricardoV94** [comment]
Not really... The change led to some scan issues that need to be addressed otherwise it would lead to a regression

**Comment by ferrine** [comment]
Marked this as draft util we figure out the scan issue

File: pymc/tests/variational/test_inference.py
Comment (ricardoV94) [inline pymc/tests/variational/test_inference.py]: Should it be an `xfail`?
```suggestion
        pytest.skip("Not Implemented Inference")
```

File: pymc/util.py
Comment (ricardoV94) [inline pymc/util.py]: This already exists here: https://github.com/pymc-devs/pymc/blob/12f4cd73bae95c42a28c2f07f6aab60071fd2ca1/pymc/aesaraf.py#L546-L550

Maybe move it to `util.py`?

File: pymc/variational/approximations.py
Comment (ricardoV94) [inline pymc/variational/approximations.py]: Better to use `at.random.integers` directly. Automatic updates are handled by `compile_pymc` anyway, which is the only thing that RandomStream really does. Not relying on it avoids issues when cloning graphs (the rng.update graph still refers to the original one) because that's invisible to tools like `clone_replace`: https://github.com/Theano/Theano/issues/1467

File: pymc/distributions/logprob.py
Comment (ricardoV94) [inline pymc/distributions/logprob.py]: This does not seem necessary anymore? Even if so, it's place wouldn't be in this module.

File: pymc/variational/opvi.py
Comment (ricardoV94) [inline pymc/variational/opvi.py]: Same applies here, if you can just do `getattr(at.random, dist_name)`, you don't have the issues with the eager `default_update`

File: pymc/variational/opvi.py
Comment (ricardoV94) [inline pymc/variational/opvi.py]: Same here

File: pymc/aesaraf.py
Comment (ricardoV94) [inline pymc/aesaraf.py]: Can we remove the whole `at_rng` and `set_rng` related utilities below from this file?

https://github.com/pymc-devs/pymc/blob/12f4cd73bae95c42a28c2f07f6aab60071fd2ca1/pymc/aesaraf.py#L872-L912

**Review by ricardoV94 — DISMISSED** [review]

**Comment by ricardoV94** [comment]
@ferrine What do you think of doing the Minibatch API breaking change for V5, and fixing the MRG later, as that is not user facing anyway.

**Comment by ferrine** [comment]
> @ferrine What do you think of doing the Minibatch API breaking change for V5, and fixing the MRG later, as that is not user facing anyway.

That makes to me, we better get back to it later, than get stuck right now. It is also reasonable to improve minibatch API and VI experience later and reduce to something working now

File: pymc/variational/approximations.py
Comment (ricardoV94) [reply pymc/variational/approximations.py]: If you need to figure out the updates (e.g, inside a Scan) you can use this utility:

https://github.com/pymc-devs/pymc/blob/4cac135cd21c24ca085cfce4f27416d061c23a5e/pymc/pytensorf.py#L1063

**Comment by ricardoV94** [comment]
Codecov complains that quite some tests are not covered now? https://github.com/pymc-devs/pymc/pull/6304/files

File: pymc/tests/variational/test_inference.py
Comment (ferrine) [reply pymc/tests/variational/test_inference.py]: This is a part of a bigger issue, all references to NotImplementedInference should be xfail

Comment (ferrine) [reply pymc/tests/variational/test_inference.py]: Basically, NotImplementedInference is something that was broken because of v3-v4 update and we know about it

File: pymc/data.py
Comment (michaelosthege) [inline pymc/data.py]: This can be placed at the top

File: pymc/data.py
Comment (michaelosthege) [inline pymc/data.py]: Same here

File: pymc/tests/test_data.py
Comment (michaelosthege) [inline pymc/tests/test_data.py]: Can be at the top

File: pymc/tests/test_data.py
Comment (michaelosthege) [inline pymc/tests/test_data.py]: Can be at the top

File: pymc/tests/test_data.py
Comment (michaelosthege) [inline pymc/tests/test_data.py]:
```suggestion
    def test_not_allowed(self):
        with pytest.raises(ValueError, match="not valid for Minibatch"):
            mb = pm.Minibatch(at.as_tensor(self.data) * 2, batch_size=20)

    def test_not_allowed2(self):
        with pytest.raises(ValueError, match="not valid for Minibatch"):
            mb = pm.Minibatch(self.data, at.as_tensor(self.data) * 2, batch_size=20)
```

File: pymc/tests/test_data.py
Comment (michaelosthege) [inline pymc/tests/test_data.py]: Which `AssertionError` is that? (Please match a part of the message..)

**Comment by michaelosthege** [comment]
I marked the VI `test_profile` tests as `XFAIL` and fixed flakyness of the Moyal logcdf test which is also currently ❌ on main.

File: pymc/aesaraf.py
Comment (michaelosthege) [reply pymc/aesaraf.py]: `set_at_rng` is called by `pm.sampling.parallel._Process._start_loop`.

In the interest of the v5 release I think we should take care of that later

**Comment by michaelosthege** [comment]
@ferrine I squashed the commits so we can rebase-merge this.

It looks like you have not set up your `@pymc-devs.org` email address in your GitHub account. That's why your pic doesn't shown on the commits..

File: pymc/tests/distributions/test_continuous.py
Comment (ricardoV94) [inline pymc/tests/distributions/test_continuous.py]:
```suggestion
        # SciPy has new (?) precision issues at {mu=-2.1, sigma=0.5, x=2.1}
```

Comment (michaelosthege) [reply pymc/tests/distributions/test_continuous.py]: it might also be that PyTensor now has smarter optimizations and therefore no longer runs into the same precision issue?

Our logcdf was [deterministic at -2227.95939](https://github.com/pymc-devs/pymc/actions/runs/3665014447/jobs/6195811961#step:7:780):

```
FAILED pymc/tests/distributions/test_continuous.py::TestMatchesScipy::test_moyal_logcdf - AssertionError: 
Arrays are not almost equal to 6 decimals
{'mu': array(2.1), 'sigma': array(0.5), 'value': array(-2.1)}
x and y -inf location mismatch:
 x: array(-2227.95939)
 y: array(-inf)
```

This value is plausible; at `sigma=0.9` it's around -50.

**Review by michaelosthege — APPROVED** [review]
🥳🚀

**Comment by ricardoV94** [comment]
@ferrine can you have one last look and merge it if you agree with the changes?

File: pymc/aesaraf.py
Comment (ricardoV94) [reply pymc/aesaraf.py]: We can, but it's only used by VI

File: pymc/tests/distributions/test_continuous.py
Comment (ricardoV94) [reply pymc/tests/distributions/test_continuous.py]: I don't think we did any stabilization improvements recently. Regardless, part of my suggestion was to also fix the typo in the original comment (as instead of has)

**Comment by ferrine** [comment]
I had one more look at the PR changes and I don't have anything to add to it. I'll merge once tests pass.

**Comment by michaelosthege** [comment]
Just fixuped the edit of the comment into the moyal logcdf commit.

Didn't want to waste CI, so I merged it right away =)

**Pull request merged**
